configs/logger/aim.yaml

```diff
@@ -0,0 +1,21 @@
+# https://aimstack.io/
+
+# example usage in lightning module:
+# https://github.com/aimhubio/aim/blob/main/examples/pytorch_lightning_track.py
+
+aim:
+  _target_: aim.pytorch_lightning.AimLogger
+  repo: ${paths.output_dir} # .aim folder will be created here
```

Comment:
I'm not familiar with Aim, but I noticed all experiments need to be saved in the same .aim folder in order to display all of them in the UI, or am I misunderstanding how it works?

If so then this doesn't go well with creating .aim folder in output_dir since its path changes on each run. Same issue as with neptune logger.

Maybe it would be better to set this to `${paths.root_dir}` as default?

Reply:
This is a very good question and is something I wanted to test out today since I've only recently started using Aim. I chose the output dir because I assumed it would work like Tensorboard where it recursively searches through folders and process/display multiple "repositories" under one UI server instance. My hope is that it will work, but I'll test it out later today, I'm just waking up :)

Reply:
Looks like you're right! You can't run `aim up --repo <directory_above_multiple_directories_each_containing_.aim>`. Aim UI can only view one repository at a time. However, a single repository can contain multiple runs. Also, Aim has support for remote tracking where you have a remote repository located on a [server](https://aimstack.readthedocs.io/en/latest/using/remote_tracking.html#) and can have multiple hosts (e.g., different compute nodes in a cluster) all writing their Runs to this single remote repository.

So I see two options here:
1. Change `aim.repo` in `aim.yaml` to `${paths.root_dir}` so that we have a single Aim repo. Multiple calls to `python src/train.py --logger=aim` will result in multiple Aim runs being created (each with a different run hash) but all of them writing to the same repository. This is fine for local Hydra runs and multiruns, but not so good for remote Hydra runs and multiruns, like if you're using the submitit launcher Hydra plugin to launch jobs. This is assuming that Aim even allows multiple runs to concurrently write to a repo without using Aim remote tracking. Another option is:

2. We can change `aim.repo` in `aim.yaml` to `${paths.root_dir}` but also provide commented-out lines underneath:

```yaml
aim:
  _target_: aim.pytorch_lightning.AimLogger
  repo: ${paths.root_dir}
# repo: "remotely/located/folder" # recommended if using Hydra multirun for launching multiple jobs on a cluster
# repo: "aim://ip_address:port" # same as above except Aim repo is now managed by an Aim remote tracking server, see https://aimstack.readthedocs.io/en/latest/using/remote_tracking.html#
```

I'll test out concurrently logging multiple runs to a single Aim repo with and without using Aim remote tracking, and I'll report back.

Reply:
> \# repo: "remotely/located/folder" # recommended if using Hydra multirun for launching multiple jobs on a cluster

I'm not sure if I understand this right so correct me if I'm wrong. 

The `paths.root_dir` doesn't work on a cluster since its a global path on user machine, which is inferred when launching the script (before sendings jobs to the cluster)?

This is why we need `remotely/located/folder`, which is a path specifically on a remote cluster, not on user local machine which they use to launch the jobs.  

I think the way to to deal with that is to change the `paths.root_dir` path in config to some location on the cluster. This can be done after the job is send to the cluster. For example, we can add a call to utility function inside the `main()` in entry file. This function could verify if we're on a cluster, and maybe inferr the `paths.root_dir` again? Or set it to work dir or environment variable.

This way any other config or code part that utilizes `paths.root_dir` will also always work when launching jobs remotely

Reply:
>The paths.root_dir doesn't work on a cluster since its a global path on user machine, which is inferred when launching the script (before sendings jobs to the cluster)?
>
> This is why we need remotely/located/folder, which is a path specifically on a remote cluster, not on user local machine which they use to launch the jobs.

Technically, `paths.root_dir` can work on a cluster just fine. However, the concern I had is when dealing with the typical cluster environment where there are multiple filesystems each optimized to be used to store specific kinds of data in specific kinds of ways (e.g., high frequency writing of small files vs. low frequency reading of large files). So for example, your code and libraries would be stored in one filesystem (let's call it `$HOME`), your datasets would be stored in another (e.g., `$DATASETS`), and the outputs of your jobs would be stored in another (e.g., `$SCRATCH`). Let's assume that the filesystem optimization is such that you're not allowed to work around it (e.g., you can't store logs, checkpoints, and saved images anywhere besides `$SCRATCH`). 

So with this setup in mind, I want to explain two scenarios and I have some questions for the first:

1. Let's say you launch your jobs from `$HOME/project_root` using `--multirun` and the submitit launcher, and you want Hydra to tell submitit to launch the jobs from a remote location, i.e., change `sweep.dir` to `${env:SCRATCH}/${task_name}/multiruns/${now:%Y-%m-%d}/${now:%H-%M-%S}`. Submitit will pickle the Hydra task function, save it to the `.submitit` folder created at `sweep.dir`, and tell the compute node to run the job using that pickled task function. Will `paths.root_dir` point to `${env:SCRATCH}/${task_name}/multiruns/${now:%Y-%m-%d}/${now:%H-%M-%S}` or will it still point to `$HOME/project_root`? If it's the latter, then there would be a problem if we decided to have `aim.repo` point to `paths.root_dir` since it would mean the job would be saving run stats to a repo created at `$HOME/project_root`, which would break the "rules" of not allowing logs being saved to `$HOME`. If it's the former, that `paths.root_dir` would be pointing to `${env:SCRATCH}/${task_name}/multiruns/${now:%Y-%m-%d}/${now:%H-%M-%S}`, then it'd be okay since Aim would be writing to the `$SCRATCH` filesystem.
2. In this scenario, we launch our jobs from `$HOME/project_root` but we _don't_ want Aim saving its run stats to `paths.root_dir` regardless of whether it's in `${env:SCRATCH}/${task_name}/multiruns/${now:%Y-%m-%d}/${now:%H-%M-%S}` or not and we _do_ want Aim savings its run stats to a Aim repo located in `$SCRATCH/aim_repos`. In this case, `paths.root_dir` wouldn't be useful nor would we want to change it to `$SCRATCH/aim_repos` as that's not where the project root is located.

It's because of these two cases I included `paths.root_dir: "remotely/located/folder"` and `repo: "aim://ip_address:port"` to imply that the repo directory does not need to be in the same filesystem/server as where the job's project root is located. Perhaps a more sustainable approach would be to have a new config in `paths.default` called `remote_data_dir` and `remote_log_dir` for user-chosen data and log directories. So for example, `remote_data_dir: $DATASETS` and `remote_log_dir: $SCRATCH/logs` or something like that. Would that make sense?

>I think the way to to deal with that is to change the paths.root_dir path in config to some location on the cluster. This can be done after the job is send to the cluster. For example, we can add a call to utility function inside the main() in entry file. This function could verify if we're on a cluster, and maybe inferr the paths.root_dir again? Or set it to work dir or environment variable.

Just a heads up that this suggestion to dynamically change `paths.root_dir` to some location on the cluster might have some downstream complications and unintended side-effects. Ideally you wouldn't want to dynamically change Hydra configurations after you've launched the jobs (specifically in `main()`) because at that point you will have to also change the same configurations in the yaml files that were saved by Hydra during compile time under the `.hydra` folder. You can go this route, but it will require also updating the [Hydra configurations that were saved to disk by Hydra during compile time](https://github.com/facebookresearch/hydra/blob/f8940600d0ab5c695961ad83abd042ffe9458caf/hydra/core/utils.py#L175-L181). I'm sure you know this already, but just bringing it up as it can get a little messy and might not be necessary for this use case.

Reply:
Regarding the scenario 2 - I get the idea of `repo: "aim://ip_address:port"`. This means we send aim experiment data to folder on remote server where we can access all of them at once, and I agree it would be great mention this in the aim config through apprioprate comment as you suggested. Sorry I didn't make this clear before.

Regarding scenario 1 - ok this is a bit complex so let me describe the order of things to make sure we're on the same page. 

We're launching the `python src/train.py`:
1. First, the `pyrootutils.setup_root()` gets executed (because it's located at the top of the `train.py` script so it runs before hydra is initialized). This call sets the `PROJECT_ROOT` environment variable based on the location of `.project-root` file on user machine (not the remote server/cluster machine)
2. Next, the `main()` function in `train.py` gets executed, which is decorated with `@hydra.main` decorator. This decorator means this is the point at which hydra gets initialized for the first time, and the job gets send to the cluster if the launcher config was attached. Sending job to the cluster means just sending the python function to be executed - this means it doesn't provide remote server with required data and any other necessary files from your local machine.

The `paths.root_dir` in cfg is by default set to `${oc.env:PROJECT_ROOT}`.

It's important to realise that `${oc.env:PROJECT_ROOT}` gets resolved only at the moment this config variable is directly accessed in the code. This resolve is not remembered (it's executed every time the config variable is accessed).

Example:
```python
@hydra.main(version_base="1.3", config_path="../configs", config_name="train.yaml")
def main(cfg: DictConfig) -> Optional[float]:
    print(cfg.paths.root_dir) # this will print the path inferred earlier, e.g. "home/user/projec_name/"

    os.environ["PROJECT_ROOT"] = "some/new/path/"

    print(cfg.paths.root_dir) # this will print "some/new/path/"
```

Now if the job was send to the cluster, then `PROJECT_ROOT` variable will not be available because it was set on local user machine. 

However, [hydra exposes ways to setup environment variables on remote servers trough config](https://hydra.cc/docs/configure_hydra/job/#hydrajobenv_copy), for example:
```yaml
hydra:
  job:
    env_copy:
      - AWS_KEY # this will copy env var from your local machine to remote environment on which jobs is executed
```

So if you want to setup root dir, data dir, and logs dir on remote server/cluster, there are many approaches, for example:
1. You could specify all remote server paths in `.env` on your local machine and make sure they're copied to remote environment with hydra config as above
2. You cold specify paths as simple strings in `- optional local: default.yaml` (this is config we have in train.py which is excluded from version cotrol)

I realize it's all confusing and I'm currently thinking about what approach should be encouraged as best practices for working with remote clusters. Feel free to share your thoughts if you have some ideas

Reply:
Going back to the PR, I think for now we can just do this:
```
# .aim folder will be created here
repo: ${paths.root_dir} 

# repo: "aim://ip_address:port" # same as above except Aim repo is now managed by an Aim remote tracking server, see https://aimstack.readthedocs.io/en/latest/using/remote_tracking.html#
```

And figure out later what's the best way to setup root on remote machine since it's something relevant for all the loggers rather then just Aim

Reply:
> I think the way to to deal with that is to change the paths.root_dir path in config to some location on the cluster. This can be done after the job is send to the cluster. For example, we can add a call to utility function inside the main() in entry file. This function could verify if we're on a cluster, and maybe inferr the paths.root_dir again? Or set it to work dir or environment variable.

Regarding this, you're right we should probably not modify configs after they're already saved to disk, so let's forget about this idea. I forgot we can just copy environment variables from local machine to remote through hydra, so this sort of approaches are not needed.

Reply:
Ah I completely forgot about:

```yaml
hydra:
  job:
    env_copy:
      - AWS_KEY # this will copy env var from your local machine to remote environment on which jobs is executed
```

> We're launching the python src/train.py:
> 
> 1. First, the pyrootutils.setup_root() gets executed (because it's located at the top of the train.py script so it runs before hydra is initialized). This call sets the PROJECT_ROOT environment variable based on the location of .project-root file on user machine (not the remote server/cluster machine)

I completely missed this, to be honest 😄 Now I'm looking into that line:
```python
pyrootutils.setup_root(__file__, indicator=".project-root", pythonpath=True)
```
and I have a much better understanding of what's going on now.

> 2. Next, the main() function in train.py gets executed, which is decorated with @hydra.main decorator. This decorator means this is the point at which hydra gets initialized for the first time, and the pickled job gets send to the cluster if the launcher config was attached.
> The paths.root_dir in cfg is by default set to ${oc.env:PROJECT_ROOT}.

Yup, gotcha, I saw that.

> It's important to realise that ${oc.env:PROJECT_ROOT} gets resolved only at the moment this config variable is directly accessed in the code. This resolve is not remembered (it's executed every time the config variable is accessed).

Yup, this much I know!

>Now if the job was send to the cluster, then PROJECT_ROOT variable will not be available because it was set on local user machine.

Gotcha 👍 

>So if you want to setup root dir, data dir, and logs dir on remote server/cluster, there are many approaches, for example:
> 1. You could specify all remote server paths in .env on your local machine and make sure they're copied to remote environment with hydra config as above
> 2. You cold specify paths as simple strings in - optional local: default.yaml (this is config we have in train.py which is excluded from version cotrol)
>
>I realize it's all confusing and I'm currently thinking about what approach should be encouraged as best practices for working with remote clusters. Feel free to share your thoughts if you have some ideas

Not confusing at all! You did a great job explaining this and I have a better idea of what's going on. I missed the line at the top of `train.py` and if I had seen that then I would've answered my own question haha. Thank you for clarifying.

Okay, soooo I think you hit the nail on the head with your first suggestion:

> 1. You could specify all remote server paths in .env on your local machine and make sure they're copied to remote environment with hydra config as above

In my opinion, I feel this is the most sustainable and safest approach because it allows for users to hide their remote server folder paths if they decide they want to have a remote location for their root, data, and logs directories. In a normal use-case scenario, those directories would be located in the same directory as the project (which is what it's currently set to), and this is completely okay since it's not revealing any sensitive information. However, once you decide to have a remote location for those directories, typically you'd want to hide that in a private `.env` file that isn't shared with others when it comes time to share your code after publishing your paper (for example). The way you have this template setup seems to already support that option. I see that option 2 achieves similar results, but the issue I have with that is that it's not environment variable-based, and I find that using environment variables for storing private information has its advantages when dealing with cluster environments, such as being able to easily copy over environment variables to job environments (as you showed).

Correct me if I'm wrong here. So an example of making use of option 1 would be like this:

`.env`
```text
# example of file for storing private and user specific environment variables, like keys or system paths
# rename it to ".env" (excluded from version control by default)
# .env is loaded by train.py automatically
# hydra allows you to reference variables in .yaml configs with special syntax: ${oc.env:MY_VAR}

# MY_VAR="/home/user/my/system/path"

PRIVATE_REMOTE_DIR="/remotely/located/folder"
```

`aim.yaml`
```yaml
# https://aimstack.io/

aim:
  _target_: aim.pytorch_lightning.AimLogger
  repo: "${oc.env:PRIVATE_REMOTE_DIR}/aim_repos" # .aim folder will be created here

  # Group runs under experiments. Takes a string value.
  # experiment: null  # 'default' if not specified.

  train_metric_prefix: "train/"
  val_metric_prefix: "val/"
  test_metric_prefix: "test/"

  # Sets the tracking interval in seconds for system usage metrics (CPU, GPU, memory, etc).
  system_tracking_interval: 10 # Set to null to disable system metrics tracking.

  # Enable/Disable logging of system params such as installed packages, git info
  # env vars, etc.
  log_system_params: true
```

`hydra/default.yaml`
```yaml
# https://hydra.cc/docs/configure_hydra/intro/

# enable color logging
defaults:
  - override hydra_logging: colorlog
  - override job_logging: colorlog

# output directory, generated dynamically on each run
run:
  dir: ${paths.log_dir}/${task_name}/runs/${now:%Y-%m-%d}_${now:%H-%M-%S}
sweep:
  dir: ${oc.env:PRIVATE_REMOTE_DIR}/my_project/${task_name}/multiruns/${now:%Y-%m-%d}_${now:%H-%M-%S}
  subdir: ${hydra.job.num}

job:
  env_copy:
    - PRIVATE_REMOTE_DIR
```

If this works then that's awesome and I'm totally on board with it. At that point, I'd suggest this for `aim.yaml`:

```yaml
# https://aimstack.io/

aim:
  _target_: aim.pytorch_lightning.AimLogger
  repo: "${paths.root_dir}" # .aim folder will be created here
  # repo: "aim://ip_address:port" # same as above except Aim repo is now managed by an Aim remote tracking server, see https://aimstack.readthedocs.io/en/latest/using/remote_tracking.html#

  # Group runs under experiments. Takes a string value.
  # experiment: null  # 'default' if not specified.

  train_metric_prefix: "train/"
  val_metric_prefix: "val/"
  test_metric_prefix: "test/"

  # Sets the tracking interval in seconds for system usage metrics (CPU, GPU, memory, etc).
  system_tracking_interval: 10 # Set to null to disable system metrics tracking.

  # Enable/Disable logging of system params such as installed packages, git info
  # env vars, etc.
  log_system_params: true
```
and then in the README maybe have a paragraph showing an example of enabling the use of remote project root, data, and log folder locations through a combination using user env vars defined in `.env`, copying it over from local to remote via `hydra.job.env_copy`, and accessing the env vars in your Hydra configs that require those paths through `${oc.env:MY_VAR}`. The paragraph can also explain why you'd want to do this, like with the use-case I provided in my earlier comment. It'd also be a nice way to show off the flexibility of this template.

How does this sound? If you like it then I'm down to help with a separate PR for that.

Reply:
> Correct me if I'm wrong here. So an example of making use of option 1 would be...

You're exactly right there.

I think there should be some ways to simplify the process. Honestly I'm not the fan of this multi-step setup. One of the problems is how to switch between working locally and working on remote machine - you don't won't to constantly change configs manually. Easiest way would be if attaching a dedicated launcher/cluster subconfig overriden the default paths: `paths.root_dir`, `paths.log_dir`, etc. 

I will need to play around a bit more to figure out how to achieve this most while maintaining the benefits of environment variables.

Reply:
>I think there should be some ways to simplify the process. Honestly I'm not the fan of this multi-step setup - feels too complicated.

I agree with you. I'm not a fan as well.

> One of the problems is how to switch between working locally and working on remote machine - you don't won't to constantly change configs manually. Easiest way would be if attaching a dedicated launcher/cluster subconfig overriden the default paths: paths.root_dir, paths.log_dir, etc.

Agreed that you wouldn't want to be constantly changing configs manually. I think having a dedicated cluster-based launcher subconfig that overrides default paths by making use of user-defined environment variables in `.env` makes sense. For example, you can have a file `paths/remote.yaml`:

```yaml
# path to remote root directory
# this requires REMOTE_PROJECT_ROOT environment variable to exist which should be created in your .env file
root_dir: ${oc.env:REMOTE_PROJECT_ROOT}

# path to remote data directory
# this requires REMOTE_DATA_DIR environment variable to exist which should be created in your .env file
data_dir: ${oc.env:REMOTE_DATA_DIR}

# path to logging directory
# this requires REOMTE_LOG_DIR environment variable to exist which should be created in your .env file
log_dir: ${oc.env:REMOTE_LOG_DIR}

# path to output directory, created dynamically by hydra
# path generation pattern is specified in `configs/hydra/default.yaml`
# use it to store all files generated during the run, like ckpts and metrics
output_dir: ${hydra:runtime.output_dir}

# path to working directory
work_dir: ${hydra:runtime.cwd}
```

Anyways, this is starting to detract from this PR. For the sake of this PR I think incorporating your suggestion of setting `repo: "${paths.root_dir}"` is just fine:

```yaml
# https://aimstack.io/

aim:
  _target_: aim.pytorch_lightning.AimLogger
  repo: "${paths.root_dir}" # .aim folder will be created here
  # repo: "aim://ip_address:port" # can instead provide IP address pointing to Aim remote tracking server which manages the repo, see https://aimstack.readthedocs.io/en/latest/using/remote_tracking.html#

  # Group runs under experiments. Takes a string value.
  # experiment: null  # 'default' if not specified.

  train_metric_prefix: "train/"
  val_metric_prefix: "val/"
  test_metric_prefix: "test/"

  # Sets the tracking interval in seconds for system usage metrics (CPU, GPU, memory, etc).
  system_tracking_interval: 10 # Set to null to disable system metrics tracking.

  # Enable/Disable logging of system params such as installed packages, git info
  # env vars, etc.
  log_system_params: true
```

I'll open up a PR in the Aim repo for fixing their AimLogger issue and will reference it here. I'll also try to incorporate explicit closing of any Aim runs within the `close_loggers()` utility function. I'll have to figure out how to pass the trainer object from `train()` up to `wrap()` so that `close_loggers()` can take it in as an argument and have access to its properties. This would be the only way to close the run as you can't access it in a global fashion like you can with wandb. I can also test to see if Aim automatically closes runs upon program exit—that would eliminate the need of any explicit closing.